app/models/user.rb

```diff
@@ -19,7 +19,17 @@ class User < ApplicationRecord
   scope :halfyear_actives, ->(time = Time.now) { logged_in_since(time - 6.month) }
   scope :active, -> { joins(:person).where(people: {closed_account: false}) }
 
-  devise :database_authenticatable, :registerable,
+  attribute :otp_secret
+
+  devise :two_factor_authenticatable,
+         otp_secret_encryption_key: "c0f2d6a8fa2873e40be5768f99b99260a3b0462ad9dd3f3\
+71432751b4e632f6ab4519b97dc7fdf5fa90c4dfdffa3ed0a917cd7a51382bac8b26166a5540bc603"
```

Comment:
Not familiar with this but usually secret keys are not available on Github ;)

Reply:
I had a look at the 2fa topic already too and compared different possible solution, and I think the `devise-two-factor` @lislis is using here was the one I would have chosen too. However this lib requires a key to encrypt the OTP secret, which I don't know if we really need this. But there is no easy way to disable encryption (no direct config flag).

The solution is the documentation is to use an env-variable for the key:

```ruby
 :otp_secret_encryption_key => ENV['YOUR_ENCRYPTION_KEY_HERE']
```

However this would add more complexity to the installation/upgrade process because every podmin would be needed to set this. So I don't like this solution.

We could add the key to the diaspora.yml, but that wouldn't solve the problem, that every podmin would need to change it (and you can't change it later, you always have to use the same key). Also podmins would need to set a key with the correct length, which again is even more complexity.

A static key would solve this problem, but I agree with you, it looks a bit strange to see this on GitHub ;)

We could try to disable the encryption with setting an `if` option and defining the attribute before the library does it, so something like this maybe works (didn't test it):

```ruby
attr_encrypted :otp_secret, if: false
```

The reason why I'm not sure if the encryption is really needed is: When somebody is already in the database and the encryption protects from reading the OTP secret, the attacker is already in the database and can access all other data and maybe even change stuff. When the attacker is on the machine where diaspora runs, they probably even have access to the key and can decrypt the secret. So the additional protection with the encryption is very low compared with the additional complexity. An attacker couldn't even do anything with the OTP secret, without the password you can't login, and you can't use the OTP secret to login on other pages, because it is a generated value only for diaspora.

A possibility would be to generate a key with the correct size at the first start (similar to the `secret_token` for the session cookie). But then we still need to make sure that podmins copy the key to a new server if they migrate the installation to a new hardware and include it in the backups, otherwise all users with 2fa enabled can't login anymore (while when they forget to migrate the session secret token only all sessions are invalid, and people just need to login again).

So I think no key (no encryption, if possible) or a static key is currently the best solution, but I'm open for other opinions.

Reply:
>However this would add more complexity to the installation/upgrade process because every podmin would be needed to set this. So I don't like this solution.

We already do that for Camo, and simply ask the podmin to add `export` in there `.bashrc`. This looks good enough to me, but I agree the encryption is maybe not needed.

Reply:
But Camo is optional for podmins, while this is optional for the users to activate, but it's not optional for podmins, so every podmin would need to do it when updating or in the future when installing diaspora. Diaspora alone currently doesn't need any env variables (all options are in the `diaspora.yml`), and the `.bashrc` wouldn't be the correct place to add something like this (it would maybe work for some situations, but it's nothing I want to write in the official installation guides).

Reply:
Don't we already do the key generation for OpenID Connect too?

Reply:
> Don't we already do the key generation for OpenID Connect too?

If it's already being done somewhere that'd be great! Otherwise I'm fine Tux's suggestion, too.

Reply:
> Don't we already do the key generation for OpenID Connect too?

Well, we kinda do, but that's happening in [`lib/api/openid_connect/id_token_config.rb`](https://github.com/diaspora/diaspora/blob/ae17248b2ecb462d7b7471d3173b27c0cb6c64a8/lib/api/openid_connect/id_token_config.rb) and isn't the nicest way of doing that, to be honest. However, there is a [task to generate the session secret](https://github.com/diaspora/diaspora/blob/ae17248b2ecb462d7b7471d3173b27c0cb6c64a8/lib/tasks/generate_session_secret.rake), and one could easily take that as a template for a new task.